pkg/server/endpoints/node/handler_test.go

```diff
@@ -1,1072 +1,1056 @@
 }
 
-func TestAuthorizeCallForCallsRequiringAgentSVID(t *testing.T) {
-	for _, method := range []string{"FetchX509SVID", "FetchJWTSVID"} {
-		testAuthorizeCallForCallsRequiringAgentSVID(t, method)
-	}
+func (s *HandlerSuite) TestFetchX509SVIDWithNoRegistrationEntries() {
+	s.attestAgent()
+	upd := s.requireFetchX509SVIDSuccess(&node.FetchX509SVIDRequest{})
+	s.assertBundlesInUpdate(upd)
 }
 
-func testAuthorizeCallForCallsRequiringAgentSVID(t *testing.T, method string) {
-	t.Logf("testing authorization for %s", method)
+func (s *HandlerSuite) TestFetchX509SVIDWithNoCSRs() {
+	s.attestAgent()
 
-	peerCert := getFakePeerCertificate()
-	peerCtx := peer.NewContext(context.Background(), getFakePeer())
+	s.createBundle(otherDomainBundle)
+	entry := s.createRegistrationEntry(&common.RegistrationEntry{
+		ParentId:      agentID,
+		SpiffeId:      workloadID,
+		FederatesWith: []string{otherDomainID},
+	})
+	upd := s.requireFetchX509SVIDSuccess(&node.FetchX509SVIDRequest{})
 
-	ds := fakedatastore.New()
-	catalog := fakeservercatalog.New()
-	catalog.SetDataStores(ds)
-	log, logHook := test.NewNullLogger()
-	handler := NewHandler(HandlerConfig{Log: log, Catalog: catalog})
+	s.Equal([]*common.RegistrationEntry{entry}, upd.RegistrationEntries)
+	s.assertBundlesInUpdate(upd, otherDomainBundle)
+	s.Empty(upd.Svids)
+}
 
-	var now time.Time
-	handler.hooks.now = func() time.Time {
-		return now
-	}
+func (s *HandlerSuite) TestFetchX509SVIDWithMalformedCSR() {
+	s.attestAgent()
 
-	fullMethod := fmt.Sprintf("/spire.api.node.Node/%s", method)
+	s.requireFetchX509SVIDFailure(&node.FetchX509SVIDRequest{
+		Csrs: [][]byte{[]byte("MALFORMED")},
+	}, codes.Unknown, "failed to sign CSRs")
+	s.assertLastLogMessageContains("failed to parse CSR")
+}
 
-	// no certificate
-	ctx, err := handler.AuthorizeCall(context.Background(), fullMethod)
-	require.EqualError(t, err, "agent SVID is required for this request")
-	require.Nil(t, ctx)
+func (s *HandlerSuite) TestFetchX509SVIDWithUnauthorizedCSR() {
+	s.attestAgent()
 
-	// no attested certificate with matching SPIFFE ID
-	ctx, err = handler.AuthorizeCall(peerCtx, fullMethod)
-	require.EqualError(t, err, "agent is not attested or no longer valid")
-	require.Equal(t, "agent spiffe://example.org/spire/agent/join_token/token is not attested", logHook.LastEntry().Message)
-	require.Nil(t, ctx)
+	s.requireFetchX509SVIDFailure(&node.FetchX509SVIDRequest{
+		Csrs: s.makeCSRs(workloadID),
+	}, codes.Unknown, "failed to sign CSRs")
+	s.assertLastLogMessageContains(`not entitled to sign CSR for "spiffe://example.org/workload"`)
+}
 
-	// good certificate
-	_, err = ds.CreateAttestedNode(context.Background(), &datastore.CreateAttestedNodeRequest{
-		Node: &common.AttestedNode{
-			SpiffeId:         "spiffe://example.org/spire/agent/join_token/token",
-			CertSerialNumber: peerCert.SerialNumber.String(),
-			CertNotAfter:     peerCert.NotAfter.Unix(),
-		},
+func (s *HandlerSuite) TestFetchX509SVIDWithAgentCSR() {
```

Comment:
We should have a case where we try to rotate the agent using an "old" svid i.e. one that doesn't match the last issued in attested node entries

Reply:
Yep, good catch. Added.